shared-libs/rules-engine/src/target-state.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * @module target-state
+ *
+ * Stores raw target-emissions in a minified, efficient, deterministic structure
+ * Handles removal of "cancelled" target emissions
+ * Aggregates target emissions into targets
+ */
+
+/** @summary state
+ * Functions in this module all accept a "state" parameter or return a "state" object. This state has the following structure:
+ * 
+ * @example 
+ * target_id: {
+ *   id: 'target_id',
+ *   type: 'count',
+ *   goal: 0,
+ *   ..
+ *   
+ *   emissions: {
+ *     emitted_id: {
+ *       requestor_id: {
+ *         pass: boolean,
+ *         date: timestamp,
+ *         order: timestamp,
+ *       },
+ *       ..
+ *     },
+ *     ..
+ *   }
+ * }
+ */
+
+module.exports = {
+  /**
+   * Builds an empty target-state.
+   *
+   * @param {Object} settingsDoc Settings document
+   * @param {Object} userDoc User's hydrated contact document
+   */
+  createEmptyState: (settingsDoc, userDoc) => {
+    const targetDefinitions = settingsDoc.tasks && settingsDoc.tasks.targets && settingsDoc.tasks.targets.items || [];
+    const isTargetInContext = targetDefinition => {
+      if (!targetDefinition.context) {
+        return true;
+      }
+
+      const context = new Function('user', `return (${targetDefinition.context});`);
```

Comment:
As discussed in Slack I think the old implementation using `$parse` is safer due to its more limited attack surface. I realise it makes the API more complicated because now you need to inject `$parse` somewhere, but I think it's worth it.